[DBM] Database Optimization (#37221)
DataDog/documentation: content/en/database_monitoring/bits_database_optimization.md

---
title: Bits Database Optimization
description: Review and implement database query optimizations identified by Bits AI.
further_reading:
- link: "/database_monitoring/"
  tag: "Documentation"
  text: "Database Monitoring"
- link: "/database_monitoring/query_metrics/"
  tag: "Documentation"
  text: "Exploring Query Metrics"
- link: "/database_monitoring/connect_dbm_and_apm/"
  tag: "Documentation"
  text: "Correlate Database Monitoring and Traces"
---

## Overview

Bits Database Optimization detects underperforming queries across your database fleet, identifies optimizations validated against a simulated copy of your environment, and delivers the result as a pull request fixing the exact code that issued the query.

Optimization candidates are selected automatically from Database Monitoring telemetry, with no additional setup required. Candidates are identified by highest potential impact, focusing on query execution times, blocking queries, and regressed queries.

> Bits Database Optimization does not require write access to your database, and does not export or use actual data from your environment. Optimizations are empirically tested against database simulations populated with synthetic data using statistical properties of your schema.

Reviewer: Let's format this as a note like we do here: https://docs.datadoghq.com/database_monitoring/setup_postgres/selfhosted/?tab=postgres15#example-agent-configurations

> 

Reviewer: We can use this screenshot to highlight the overall feature: https://drive.google.com/file/d/1zh8C4E_1fy3D3Tvn-aXwYDoajfULoaJP/view?usp=sharing

> ## Prerequisites

Reviewer: Should we format this as a table like we do here? 

https://docs.datadoghq.com/database_monitoring/setup_postgres/selfhosted/?tab=postgres15#configure-postgres-settings

Author: Are you OK keeping as bullets? The table works great for required config settings (i.e. param > value > description), but we don't need those here. I think the bullets are better-suited for a more straightforward list like this that's just required services.

Reviewer: Work for me!

- **Database Monitoring** is configured for the target database instances. See [Database Monitoring Setup][1].
- **Schema collection** is enabled on the target instances.
- For automated PR creation:

>     - **APM** must be configured for the services that issue the queries you want to address. See [Correlate Database Monitoring and Traces][2] for more information.

Reviewer: @amw-zero is this true?

    - A **GitHub repository** must be linked in your Datadog organization.

> ## Reviewing and implementing database optimizations

Reviewer: Can we change this to Viewing Optimizations and then have subsections for Queries List, Optimization panel?

And then lets create a separate section just for the PR piece.

Author: Thanks for this - this structure works much better.

1. On the [Database Monitoring > Queries][3] screen, queries with optimizations available have an AI icon in the Status column. Hover over an icon to see a summary of the optimization, and click the icon to open the Optimization panel.
    - To filter the query list by optimization type, select an option from **Optimizations** above the list.

> {{< img src="database_monitoring/database_optimization_queries.png" alt="The status column on the Queries screen, showing AI icons in query rows where an optimizations are available. " style="width:100%;">}}

Reviewer: We can use this link for the image https://drive.google.com/file/d/1Xa3eZS3AY52XaZOHsjsRW2lGBSNE_BZf/view?usp=sharing

2. The Optimization panel includes a summary of the query issue, the optimized query used in the simulation, and a visualization of the Simulated Performance Impact.
3. Explore the Simulated Performance Impact visualization for more details about improvements:
    - Hover over the improvement summary (for example, "3.7x More Efficient") to view before-and-after execution times, logical reads, and shared blocks dirtied. The table shows the average, median, P95, and maximum for each metric.
    - Hover over each item in the visualization to view more details.

> {{< img src="database_monitoring/database_optimization_simulated_performance_impact.png" alt="An example Simulated Performance Impact visualization, showing a query optimized to 44.7x more efficient and 92x faster." style="width:100%;">}}

Reviewer: You can use this image for the simulation map 

https://drive.google.com/file/d/14fLYNYIyuZPtg-ZQEkceXazbiJuQjzIf/view?usp=sharing

4. Click **Compare Plans** to view side-by-side comparisons of the current and optimized execution plans:
    - **List View** shows a hierarchical list of the execution plan's operations, with node cost and row estimates for each step.
    - **Map View** shows a visual representation of the execution plan, with the option to compare the plans by different metrics.
    - **Raw** shows the raw execution plan output.

> {{< img src="database_monitoring/database_optimization_plan_comparison_map_view.png" alt="The Compare Plans Map View, showing added and removed operations for an optimized query. " style="width:100%;">}}

Reviewer: You can use this image for plan comparison https://drive.google.com/file/d/1TZTTXVWLfb_NWb014PVF4QIrAgRk99KA/view?usp=sharing

5. To create a pull request to apply the optimization fix to your database, select **Review PR by Bits AI**. The GitHub PR opens with a pre-populated description that includes the simulation results.
<div class="alert alert-info">Automated pull requests require APM configured for the service issuing the query, and a GitHub repository linked to your Datadog organization.</div>

> ## Optimization types

Reviewer: Let's remove this section entirely because we're going to provide optimizations beyond this list as well that are entirely dynamic so we won't be able to list them all.

### Large ORDER BY without LIMIT

Sorting a large result set without a `LIMIT` forces the database to sort every matching row, often spilling to disk. Bits Database Optimization adds a `LIMIT` clause to bound the sort.

### Idle in transaction

Connections that hold an open transaction without performing work block other queries and hold locks longer than necessary. Bits Database Optimization reports the total idle time and the total blocked wait time caused by these connections, helping you understand the downstream impact.

## Priority levels

Optimizations are ranked by priority to help you focus on the highest-impact issues first:

| Priority | Optimization types | Reason |
|---|---|---|
| High | OFFSET without ORDER BY, Idle in Transaction | Data correctness risk or blocking behavior |
| Medium | Missing Index, Large ORDER BY without LIMIT | Significant performance impact |
| Low | SELECT \*, Query Rewrite | Performance suggestions |

<!-- [PLACEHOLDER: Screenshot of the PR in GitHub showing the embedded simulation summary and the diff] -->
<!-- [PLACEHOLDER: Screenshot of the "View PR" link from the normalized query view in DBM] -->

> ## Slack notifications

Reviewer: Let's remove this section as well because this isn't supported yet

---
title: Bits Database Optimization
description: Review and implement database query optimizations identified by Bits AI.
further_reading:
- link: "/database_monitoring/"
  tag: "Documentation"
  text: "Database Monitoring"
- link: "/database_monitoring/query_metrics/"
  tag: "Documentation"
  text: "Exploring Query Metrics"
- link: "/database_monitoring/connect_dbm_and_apm/"
  tag: "Documentation"
  text: "Correlate Database Monitoring and Traces"
---

## Overview

> Bits Database Optimization detects underperforming queries across your database fleet, identifies optimizations validated against a simulated copy of your environment, and delivers the result as a pull request fixing the exact code that issued the query.

Reviewer 2: ```suggestion
Bits Database Optimization detects underperforming queries across your database fleet, identifies optimizations validated against a simulated copy of your environment, and delivers the result as a pull request fixing the exact code that triggered the query.
```
i think this is a great opening paragraph but wondering if this word fits better for what's happening?

Reviewer 2: also i think the sentence length is fine, but if you want it to be 25 words or less to make Vale happy, could just do: 

```suggestion
Bits Database Optimization detects underperforming queries across your database fleet and identifies optimizations validated against a simulated copy of your environment. Then, it delivers the result as a pull request fixing the exact code that triggered the query.
```

Optimization candidates are selected automatically from Database Monitoring telemetry, with no additional setup required. Candidates are identified by highest potential impact, focusing on query execution times, blocking queries, and regressed queries.

<div class="alert alert-info">Bits Database Optimization does not require write access to your database, and does not export or use actual data from your environment. Optimizations are empirically tested against database simulations populated with synthetic data using statistical properties of your schema.</div>

{{< img src="database_monitoring/database_optimization_panel_overview.png" alt="Am optimized query in the optimization panel, showing a detailed summary of the issue, a diff for the optimized query, and a button to create a PR." style="width:100%;">}}

## Prerequisites

> - **Database Monitoring** is configured for the target database instances. See [Database Monitoring Setup][1].

Reviewer 2: was wondering if we should send people here with this link (instead of the Postgres page): https://docs.datadoghq.com/database_monitoring/architecture/

- **Schema collection** is enabled on the target instances.
- For automated PR creation:
    - **APM** must be configured for the services that issue the queries you want to address. See [Correlate Database Monitoring and Traces][2] for more information.
    - A **GitHub repository** must be linked in your Datadog organization.

## Viewing optimizations

### Query list

> On the [Database Monitoring > Queries][3] screen, queries with optimizations available have an AI icon in the Status column. Hover over an icon to see a summary of the optimization, and click the icon to open the Optimization panel.

Reviewer 2: maybe?
```suggestion
On the [Database Monitoring > Queries][3] screen, queries with optimizations available have an AI icon in the Status column. Hover over an icon to see a summary of the optimization, then click the icon to open the Optimization panel.
```

To filter the query list by optimization type, select an option from **Optimizations** above the list.

{{< img src="database_monitoring/database_optimization_queries.png" alt="The status column on the Queries screen, showing AI icons in query rows where optimizations are available." style="width:100%;">}}

### Optimization panel

The Optimization panel includes a summary of the query issue, the optimized query used in the simulation, and a visualization of the Simulated Performance Impact.

Explore the Simulated Performance Impact visualization for more details about improvements:

>   - Hover over the improvement summary (for example, "44.7x more efficient") to view before-and-after execution times, logical reads, and shared blocks dirtied. The table shows the average, median, P95, and maximum for each metric.

Reviewer: Changed this wording to now say X faster e.g. in the screenshot example it's 96.9x faster so can we update that?

  - Hover over each item in the visualization to view more details.

{{< img src="database_monitoring/database_optimization_simulated_performance_impact.png" alt="An example Simulated Performance Impact visualization, showing a query optimized to 44.7x more efficient and 92x faster." style="width:100%;">}}

Click **Compare Plans** to view side-by-side comparisons of the current and optimized execution plans:
  - **List View** shows a hierarchical list of the execution plan's operations, with node cost and row estimates for each step.
  - **Map View** shows a visual representation of the execution plan, with the option to compare the plans by different metrics.
  - **Raw** shows the raw execution plan output.

{{< img src="database_monitoring/database_optimization_plan_comparison_map_view.png" alt="The Compare Plans Map View, showing added and removed operations for an optimized query." style="width:100%;">}}

### Review the pull request

To review the PR for the optimization fix to your database, select **Review PR by Bits AI**. The GitHub PR opens with a pre-populated description that includes the simulation results.

> <div class="alert alert-info">Automated pull requests require APM configured for the service issuing the query, and a GitHub repository linked to your Datadog organization.</div>

Reviewer 2: ```suggestion
<div class="alert alert-info">Automated pull requests require APM to be configured for the service issuing the query, and a GitHub repository linked to your Datadog organization.</div>
```

  - Hover over the improvement summary (for example, "44.7x more efficient") to view before-and-after execution times, logical reads, and shared blocks dirtied. The table shows the average, median, P95, and maximum for each metric.
  - Hover over each item in the visualization to view more details.

{{< img src="database_monitoring/database_optimization_simulated_performance_impact.png" alt="An example Simulated Performance Impact visualization, showing a query optimized to 44.7x more efficient and 92x faster." style="width:100%;">}}

Click **Compare Plans** to view side-by-side comparisons of the current and optimized execution plans:
  - **List View** shows a hierarchical list of the execution plan's operations, with node cost and row estimates for each step.
  - **Map View** shows a visual representation of the execution plan, with the option to compare the plans by different metrics.
  - **Raw** shows the raw execution plan output.

{{< img src="database_monitoring/database_optimization_plan_comparison_map_view.png" alt="The Compare Plans Map View, showing added and removed operations for an optimized query." style="width:100%;">}}

> ### Create a pull request

Reviewer: ```suggestion
### Review pull request
```

> To create a PR to apply the optimization fix to your database, select **Review PR by Bits AI**. The GitHub PR opens with a pre-populated description that includes the simulation results.

Reviewer: ```suggestion
To review the PR for the optimization fix to your database, select **Review PR by Bits AI**. The GitHub PR opens with a pre-populated description that includes the simulation results.
```

<div class="alert alert-info">Automated pull requests require APM configured for the service issuing the query, and a GitHub repository linked to your Datadog organization.</div>

> ## Priority levels

Reviewer: Can you remove the Priority Levels section entirely? 

How prioritization works is still in flux so we don't want to document this for now.
